### ISSUE TITLE
A special max-SMT problem: maximizing the number of domain constraints

### ISSUE THREAD

**Issue by Adorf** [Z3Prover/z3#473]
I have some questions concerning the capabilities of Z3's tactics and tacticals.
# The problem

The problems for which we are using SMT technology have the following characteristic:

(1) There is a base SMT-problem (CSP) that contains all the obligatory constraints that must be obeyed. The base CSP is guaranteed to be sat (unless there is an error somewhere). Thus when solving such a CSP, we are not interested in answering the question whether it is sat or unsat or unknown, but we want to obtain a solution (model) for the CSP.

(2) In addtion to generating a base solution we are generating several extra solutions. For each variable (Boolean, Int, Real, ...) occuring in the CSP we have a set of special values that we would like to squeeze in into one of these extra solutions. Of course, not every special value is compatible with the CSP's constraints. Some special values are impossible. Some special values are incompatible with other special values, and therefore cannot occur in the same extra solution.

(3) We want to minimize the number of extra solutions. Thus we have to solve an optimization problem. The minimum number of solutions that we are looking for does not have to be a mathematically strict absolute minimum. An approximate minimum is sufficient for our purposes.
# Algorithm

Using a standard SMT-solver we create the desired CSP-solutions in the following way:

The algorithm works with several runs, one extra solution per run. A run is started when there is at least one variable (left) that has an unused special value.

(a) Via check-sat and get-model  a base solution is computed. The solution is read out and saved in a cache. The solver context is pushed.

(b) The next variable with at least one unused special value is selected. If there is no such variable left, the cached solution is written to disk, and the run terminates.

(c) One of the variable's special values is selected. The value is assigned to the variable using a normal assert. This is a voluntary constraint that is present in addition to the obligatory constraints.

(d) A check-sat is perfomed. 
- In case of sat, the current solution is read out via get-model and is saved in the cache. The special value is marked as used and is removed from the variable's set of special values. The solver context is pushed. 
- In case of unsat the context is popped. (We have no experience with the result unknown yet.) When the variable is the first variable in the current run, the special value is removed, because it is impossible. Otherwise it is put back into the variable's set of special values. Then we continue with step (b) .

The algorithm is deterministic and is guaranteed to terminate.

As an optimization, for each variable we also mark a special value as used if it accidentally occurs in a solution.
# Remarks

Obviously, our problem is similar to a max-sat problem, except that we have two sets of constraints, namely one set with obligatory constraints, and one set with voluntary constraints. The former are static and stay, whereas the latter are dynamic, and may be present and fulfilled or not.
# Questions

Here are my questions:

(1) is it possible to internalize the deterministic algorithm above using Z3's current set of tactics and tacticals? 

(2) If not, what would be missing? Would it be possible to conceive one or more tactics or tacticals that, in cooperation with tactics and tacticals already existing in Z3, would solve the above optimization problem? In my view, it would be sufficient to internalize a single run into Z3.

(3) Could a stochastic local search (SLS) algorithm be used for the optimization? I am thinking of equating an extra Boolean variable with each of the extra equations. The optimization would mean to maximize in each run the number of these extra Boolean variables that are true. Over time the number of Boolean variables that need to be considered would diminish from run to run, until none is left over.

**Comment by Adorf** [comment]
Admittedly my description and the questions above is rather complex and long-winded. I give it another try:

The problem
==========
(1) Given a base SMT-problem (CSP), which is guaranteed to be sat. Thus, when solving the base CSP, we obtain a solution (model), called base solution.

(2) Given a set of variable-value pairs. We wish to generate a set of extra solutions that comprise all variable-value pairs compatible with the base CSP. The cardinality of the set shall be minimized. The minimum does not have to be exact. Rather we are satisfied with an approximate minimum (in order to lower the time and effort for producing the extra solutions).

An algorithm (simpler than the one described above)
==========================================

Phase A: Weed out incompatible variable-value pairs
-------------------------------------------------------------------
(1) Compute a base solution and save it.

(2) For each variable-value pair: 
- push the context
- use assert to assign the value to the variable
- check-sat. 
- If unsat, discard the variable-value pair. 
- pop the context.

The set of remaining variable-value pairs are all compatible with the base CSP.

Phase B: Generate extra solutions
-------------------------------------------
Compute a base solution.
While there is an unused variable-value pair
- push the context
- add  to the base CSP all unused variable-value pairs via (assert (= var val))
- compute a solution that maximizes the number of variable-value pairs occuring in the solution and save it
- pop the context
- remove all variable-value pairs that occured in the solution

Basically each round in the while loop executes a max-sat algorithm with two sets of constraints:
- a base set of constraints that must be obeyed
- a second set of simple variable-value constraints of which a maximum number shall be obeyed

Remarks
------------
By splitting the algorithm into two phases, the algorithm for the second phase can be much simpler, and there are more possibilities for implementing it.

I suspect that a general weighted max-sat algorithm would work, if the weights assigned to the base constraints were sufficiently high.

**Comment by NikolajBjorner** [comment]
I have two readings of your question:

1. There is no real minimization in step 2, but rather you have a case of the form:
Given a CSP formula phi[x1,..,xn, y] and domain constraints x1 in D1, .., xn in Dn, where D1, ,.., Dn are finite set of values you want to compute the set of tuples (d1, d2, .., dn), such that d1 in D1, d2 in D2, .., dn in Dn, where phi[d1, ..,, dn, y] is satisfiable.
This is classical all-SAT or let's call it all-SMT for finite domains. If you keep blocking old solutions in a naive way Z3 will eventually suffer from memory overhead.

2. You wish to compute the maximal subset of x1, .., xn, such that the constraints x_i in D_i are true and also phi[x1,.., xn, y] is true for some y. This is a Max-SAT/Max-SMT problem. You can formulate this directly using "soft constraints". The tutorial http://rise4fun.com/z3opt elaborates on the optimization features in Z3. You basically assert phi, and you create formulas corresponding to "x_i in D_i" (a disjunction of equalities) and assert them as soft constraints.

**Comment by Adorf** [comment]
@NikolajBjorner: let me follow your example and use a mathematically precise language to concisely state our problem:

Problem description
-------------------------
- There is a quantifier-free background (base) CSP y, which is known to be sat. (Since y may contain nonlinearities, we cannot rule out the possibility that the SMT-solver returns unknown, but that is a different story.)
- All I variables x1, ..., xI in y have finite domains D1, ..., DI. The necessary domain constraints are part of the background CSP y.

- For each variable xi there is a finite sequence of Ji finite subdomains Sij.
- For each i = 1, ..., I and each j = 1, ..., Ji the subdomain Sij is a subset of the domain Di.
- For each i the Ji subdomains Sij are disjoint.
- For each i and j there is a domain constraint cij, which restricts variable xi to the subdomain Sij.

For simplicity, we assume that each of the domain constraints cij individually is compatible with the background CSP y, i.e. the CSP obtained by adding a single cij to y is still sat.

- We wish to obtain k = 1, ..., K solutions (models) Mk.
- For all i and j there shall be at least one solution Mk, in which variable xi has a value from subdomain Sij.
- The number K of solutions shall be minimized.

Clearly, for the number K of solutions we have the inequalities K >= Ji for all i, since the Sij are disjoint for fixed i.

If a mathematically strict minimum for K were too costly to compute, we would be satisfied with an approximate minimum.

Heuristic algorithm (A)
----------------------------
Our approximate, heuristic algorithm (A) described in the earlier comments works like this:

```
;; initialization phase, for now outside the solver
for i = 1 to I
  put all constraints cij into a FIFO queue Ci
endfor

enter y to the solver
check-sat ; we are at context stack level 0 and have a base solution for y
push the context ; preserve the solver's internal sat state

;; computation phase
while there is a variable xi with non-empty queue Ci
  for i = 1 to I 
    if Qi is non-empty
       pop a domain constraint cij from queue Ci
	   assert cij
	   check-sat
       if sat
	     push the context ; preserve the solver's internal sat state
	   else
	     pop the context ; discard the solver's internal unsat state
	     add cij at the end of Ci
	   endif
    endif
  endfor
  get-model and save it externally
  pop sufficiently often so that we are back at context stack level 0 where we have preserved our base sat context for y
endwhile
```

In effect, each variable gets one chance to acquire a value from a subdomain Sij via the domain constraint cij. If cij is incompatible with the domain constraints previously issued in the current round, the variable xi is left free. Its value will be chosen by the solver.

The push and pop machinery working on the solver's internal context stack in our experience enormously improves effiency.

Several variations/improvements are conceivable:

The algorithm (A) sketched above simply uses the numbering of the variables, and it ignores the size of the queues. A probable improvement would be to use a priority queue for X, and to sort the variables according to the size of the associated constraint queues Ci.

Heuristic algorithm (B)
----------------------------
In heuristic algorithm (B) the computation phase would be altered into:

```
form a variable priority queue X which initially contains all variables xi with a non-empty constraint queue Ci ; all variables in our case
sort the variables xi in X according to the size of the associated constraint queue Ci, in descending order
while the variable queue X is non-empty
  pop a variable xi from the variable queue X 
  pop a domain constraint cij from the constraint queue Ci
  assert cij
  check-sat
  if sat
    push the context onto the context stack ; for later use
  else
    pop the context stack ; the context has become useless
    add cij at the end of Ci
  endif
  if the variable queue X is empty
     get-model and save it externally
     pop the context stack sufficiently often so that we are back at context stack level 0 where we have preserved our base sat context for y
  	 add all variables xi with non-empty constraint queue Ci back into variable queue X
  endif
endwhile
```

Clearly, other algorithms for sorting the variable queue X come to mind, such as most constrained first, etc. Such a sort algorithm would require an analyze of the constraints in the base CSP y. If the algorithm ran completely internally inside the solver, the solver would presumabmly employ its own prioritization strategy for the variables.

The next improvement that comes to mind is to not assert the remaining domain constraints cij
one at a time. Instead, when variable xi is considered, the disjunction of the remaining domain constraints cij is asserted. This gives the solver more latitude to satisfy one of the domain constraints cij.

Heuristic algorithm (C)
----------------------------
In heuristic algorithm (C), instead of constraint queues Ci, we would use contraint sets Ci. The computation part would read:

```
form a variable queue X which contains all variables xi with a non-empty constraint set Ci
sort the variables xi in X according to the size of the associated constraint set Ci, in descending order
while the variable queue X is non-empty
  pop a variable xi from the variable queue X 
  assert the disjunction of all constraints cij in Ci, i.e. (or ci1 ci2 ...)
  check-sat
  if sat
    push the context onto the context stack ; for later use
	find out which of the mutually exclusive domain constraints cij in Ci was satisfied, and remove it from Ci
  else
    pop the context stack ; the context has become useless
  endif
  if the variable queue X is empty
     get-model and save it externally
     pop the context stack sufficiently often so that we arrive at context stack level 0 where we have preserved our base sat state for y
  	 add all variables xi with non-empty constraint set Ci back into variable queue X, and sort the latter
  endif
endwhile
```

Concurrency
----------------
There is room for concurrency. 

Let us assert the disjunctions Di = (or ci1 ci2 ...) upfront directly after defining all variables, but before asserting any constraints from the background CSP y.

```
...
(assert (or c11 c12 ... ))
(assert (or c21 c22 ... ))
(assert (or c31 c32 ... ))
...
```

Then the strategy

`(check-sat-using (par-do split-clause smt))`

would solve [C1| dressed CSPs in parallel. Here I have used the hypothetical tactical par-do (https://github.com/Z3Prover/z3/issues/479) that first applies the tactic split-clause to the original goal, and then the tactic smt concurrently to each of the subgoals produced by split-clause.

The tactical `par-do` should have a parameter, say `:max-workers`, by which one could limit the number of threads concurrently working on the subgoals. So we would issue something like

`(check-sat-using ((! par-do :max-workers 6) split-clause smt))`

which should work fine on mylaptop which is sporting 8 cores. (If, behind the scenes, Z3 supported truly distributed computing, one could use a much higher number of workers, of course.)

A problem with the strategy above is that a solution for y would be rediscovered over and over again by the different threads.

So my questions: 

1. Assuming that there were a tactical `par-do`, is there a way, using tactics and tacticals, to compute a base solution for the naked CSP y that could be re-used (via the internal context stack) when concurrently computing the dressed solutions for y plus some extra domain constraints cij?
2. How much of the algorithms above are amenable to be encoded into tactics and tacticals (including, if necessary, new hypothetical ones)?

I stop with this questions and will address the desired optimization in another comment.

Side remark
---------------
A side remark on the types we use, which is not directly relevant to the specification of the general problem:
- The variables have native types (sorts) Bool, Int, Real or String.
- Some Int variables have a 0-1 domain. We work with pseudo-booelan functions.
- All Real variables have decimal types, which is enforced by special constraints on each variable.
- (If there were subtypes in the SMT-LIB 2 world, we could define the 0-1 variables and the decimal veriables elegantly using subtypes.)
- We would like to work with String variables with regexp-xonstraints, but do not yet know how to do this within Z3.

The number Ji of subdomains typically varies from 1 to 15.

**Comment by NikolajBjorner** [comment]
Got it. The heuristic algorithms are in line with standard techniques for finding maximal satisfying subsets. However, I would also consider the facilities that let you assert with assumptions. 
Assumptions are temporary assertions that are retraced after search. They can be used instead of push/pop. The main tradeoff is as follows:
1. assumptions have to be propositional variables or negations of propositional variables. They cannot be compound expressions. To encode a compound formula F you add an assertion (a <=> F), and then assume a. 
2. Auxiliary clauses that are independent of assumptions and accumulated during search are retained. This can save on incremental calls. Push/pop only saves on pre-processing.

You can start out as follows:

            assert qij => cij  for each i, j
            assert pi <=> \/ qij  for each i

Then you call 

             check-with-assumptions p1, ..., pn

If the result is unsat, it means that it is not possible to cover all sets simultaneously.
The call returns an unsat core (not necessarily minimal), which you can use to trim the assumptions.
For example, assume that p1 is in the core, then you can perform a next call of the form:

             check-with-assumptions p2, p3, ..., pn

Eventually an assignment produces some subset of qij. Call this subset M (it corresponds to a model).

           assert \/_{qij in M} !qij  

Make another call:

            check-with-assumptions p1, ..., pn

The next model will include some qij which was not part of the previous maximal solution.

You can even break symmetries by adding some uncovered qij to the initial assumptions, but
obviously you don't want to remove qij from any cores you may encounter.

The algorithm so far accumulates a set of cores, trims them and throws them away after use.
It is possible to be more clever about the cores by relaxing the problem.
Suppose that p1, p2, p3, p4 is a core. So let us look for a solution where we are allowed to satisfy 3 of these 4 constraints. One encoding is as follows:

             p1' <=> p1 \/ p2
             p2' <=> p3 \/ (p1 & p2)
             p3' <=> p4 \/ ((p1 & p2) & p3)

which expands into:

             p1' <=> p1 \/ p2
             r1 <=> (p1 & p2)
             p2' <=> p3 \/ r1
             r2 <=> (r1 & p3)
             p3' <=> p4 \/ r2

Suppose that there is a solution to p1', p2', p3'. It is easy to see that it corresponds to a solution that satisfies 3 of the 4 constraints (this is in a nutshell the encoding suggested by Narodytska and Bacchus in AAAI 2014, used in the Eva solver and also used in Z3's maxsat implementation, see the IJCAI 2015 paper on my home page).

The next sat call is then

          check-with-assumptions p1', p2', p3', p5, p6, .., pn      

If you use the "trick" of adding a qij as an assumption you have to be careful with this transformation:
If qij is part of the core, you want to relax the other predicates but not qij. However, when you are trying to find a model that includes a different q_kl, then the relaxation should be discarded. 
So it is really not clear if the relaxation trick is preferable when the core includes qij, I would try the two versions. 

You could of course try an algorithm with assumptions that iterates over the qij (similar to your algorithms that maintain queues of uncovered cij).

**Comment by Adorf** [comment]
@NikolajBjorner: I am extremely grateful for your quick and extensive responses. 

I understand now that in Z3 the push-pop machinery only saves on pre-processing, 
but the internal state is not preserved.  Our positive experience with the push-pop machinery stemmed from working with Yices 1.

In the distant past I have worked in the planning domain, where I used the CPLEX library 
for optimizing linear goal functions. A major breakthrough occurred, when we reused a previous solution (which could be externalized and read back in) as a starting point for further computations. 
The incremental search was much faster than searching from scratch. We beneficially used this facility to add further constraints. It also worked well when we replaced constraints with new ones 
that possibly were contradictory to the original CSP.

That brings me to a question: can we get Z3 to dump the current internal state, and can it
read such a state back in? Can Z3 perform an incremental search in the way described above?

Let me elaborate on your first comment where you suggest to use soft constraints by using the Z3's ability to maximize an integer valued goal function, a capability that I had been unaware of.  Suppose we equate propositional variables pi with a disjunction of the domain constraints cij, and also equate propositional variables pi with constraints on bit (0-1) variables bi.

Using the following code we could compute a single solution (model) Mk that maximizes the number of domain constraints that are satisfied.

```
(assert y)

(declare-const pi Bool) ; for all i
(declare-const bi Int)  ; for all i
(assert (or (= bi 0) (= bi 1))) ; for all i, restricts bi to be a bit variable
(declare-const z Int)

(assert (= pi (= bi 1))) ; for all i
(assert (= z (+ b1 b2 ...)))
(maximize z)

(push)
(assert (=> pi (or ci1 ci2 ...))) ; for all i
(check-sat)
(get-model)
(pop)
```

We will obtain a sat solution (model) Mk, which is guaranteed to exist, because the background CSP y is known to be sat. Moreover, we can say that z >= 1, since there is at least one domain constraint cij that can be satisfied in addition to the obligatory background CSP y.

We would proceed as follows:
 
```
save the solution Mk away
inspect Mk, in order to find the domain constraints cij that are satisfied in Mk
remove these domain constraints cij from consideration
start another round (beginning with push and ending with pop), until no domain constraints cij are left
```

This algorithm is an elegant way for truly maximizing the number of domain constraints 
that are satisfied in each solution Mk. Nevertheless, it remains a heuristic when it comes to minimizing the total number K of solutions that are covering all domain constraints cij. I see no way how to express such a minimization in Z3's framework. Do you?

But, as I said before, it is not extremely important for us to truly minimize K. It is much more important to *quickly* arrive at a set of solutions Mk that (A) cover all cij and (B) whose number K is approximately minimal.

Our experience shows that for smaller CSPs y the number K of solutions equals the cardinality 
|Ci| of the largest set Ci comprising the domain constraints cij. So the algorithm above might be improved by introducing weights wi and multiplying each bit-variable bi with wi. High weigths wi would be associated with large constraint sets Ci

`(assert (= z (+ (* w1 b1) (* w2 b2) ...)))`

However, this approach would lead us into the domain of non-linearity with all the pitfalls associated with non-decidability. In any case, would Z3 allow us to enter such a non-linear optimization problem, and could its optimizer in principle work with a non-linear goal?

Again, it would not be a major hindrance for us, if only the linear optimization sketched above would work. And if the maximization of of the number of domain constraints fulfilled per solution takes too much time, we can leave it, and resort to one of the heuristics described before.

Once more I have used the push-pop mechanism, since I am familiar with it. If we could somehow 
preserve the internal state of Z3 and reuse it for the next round all the better. 

I still have to think about whether and how check-with-assumptions could be used for our purposes.

**Comment by Adorf** [comment]
I had some time now to think about how to possibly use Z3's command `check-with-assumptions`.

Below is the best algorithm I have found so far. pseudo-code is executed externally to the solver, whereas Lisp-style code is submitted to the solver.

```
externally define the constraint sets Ci comprising domain constraints cij
form a set C containing all constraint sets Ci

;; solve the background CSP y
(assert y)  ; enter y
(check-sat) ; compute a base solution as a starting point for further ado

set round counter k to 1
while C is non-empty
  ;; start new round k trying to produce a solution Mk
  (declare-const pik Bool) ; for current round k and all i with non-empty Ci in C
  (assert (=> pik (or ci1 ci2 ...))) ; for current round k and all i with non-empty Ci in C
  (check-with-assumptions p1k p2k ...)
  if sat
    (get-model)
    save solution Mk
    analyze solution and remove satisfied constraints cik from the sets Ci
    restore full set C with all remaining non-empty constraint sets Ci
    increment k
  else
    if possible minimize the unsat core
    analyze the unsat-core and find conflicting variables pik (k is fixed, i varies)
    temporarily remove conflicting Ci's from C, there by reducing the number of active pik
  endif
endwhile
```

If an attempt to find a solution fails in round k, it should succeed in the next round k+1, since we have removed the conflicts.

I do not know whether I have fully grasped your proposal on how to use check-with-assumptions, 
but to me it looks as if the algorithm above should work. Moreover, I believe it would be faster
than trying to truly optimize the number of domain constraints covered in each Mk. 
This believe stems from our experience (with Yices 1) that conflicts are detected much faster 
than what it takes to produce a solution

**Comment by NikolajBjorner** [comment]
closing as lack of traction
